### ISSUE TITLE
Lack of documentation

### ISSUE THREAD

**Issue by KES777** [garu/Data-Printer#99]
Please clear what does this mean: `As of version 0.13, you may also use the '-class' filter, which will be called for all non-perl types (objects).`

I may suppose you mean [this](https://metacpan.org/pod/Data::Printer#MAKING-YOUR-CLASSES-DDP-AWARE-WITHOUT-ADDING-ANY-DEPS) but there is no description for `-class`

**Comment by garu** [comment]
Hi! Thank you for using Data::Printer and for taking the time to write this, I really appreciate it.

That bit of the documentation is describing how to create custom filters. Let's review it:

---------
*Data::Printer offers you the ability to use filters to override any kind of data display. The filters are placed on a hash, where keys are the types - or class names - and values are anonymous subs that receive two arguments: the item itself as first parameter, and the properties hashref (in case your filter wants to read from it). This lets you quickly override the way Data::Printer handles and displays data types and, in particular, objects.*

```perl
use Data::Printer filters => {
  'DateTime'      => sub { $_[0]->ymd },
  'HTTP::Request' => sub { $_[0]->uri },
};
```

*Perl types are named as ref calls them: SCALAR, ARRAY, HASH, REF, CODE, Regexp and GLOB. As for objects, just use the class' name, as shown above.*

*As of version 0.13, you may also use the '-class' filter, which will be called for all non-perl types (objects).*

-------

So what I mean is, by using '-class' as a key (instead of the object's class) your code filter code will be called for every unfiltered object being parsed. If you return an undefined value, it will assume your filter was **not** successful.

So, for example:

```perl
use Data::Printer filters => {
  '-class' => sub {
    my ($object, $properties) = @_;
    warn "I was called to handle " . ref($object) . "\n";
    if ($object->isa('My::Random::Class')) {
      return "a neat object";
    } else {
       # not the object I want. Let another filter try to handle it!
      return;
    }
  },
};
```

Now, if later we do (using 5.10+ package notation):

```perl
package My::Random::Class  { sub new { bless {}, shift; } };
package Some::Other::Class { sub new { bless {}, shift; } };

my $num  = 3.14;
my $obj1 = My::Random::Class->new;
my $obj2 = Some::Other::Class->new;
p $num;
p $obj1;
p $obj2;
```

This is what we'll see as output:
```
3.14
I was called to handle My::Random::Class
a neat object
I was called to handle Some::Other::Class
Some::Other::Class  {
    public methods (1) : new
    private methods (0)
    internals: {}
}
```

So, as you can see, the custom filter was called for both objects, but NOT for the scalar. In `$obj1` it returned *"a neat object"* and thus that's what Data::Printer rendered as a dump for that object. In `$obj2`, it returned an undefined value so Data::Printer kept going, calling other filters. Since there were none, it defaulted to the general class handling filter. You could do this to render a different output for all classes that `->can('stringify')` for example, or that inherit from a common parent, or anything you want, really. Another example would be if you were dumping an object which held a very large string, then instead of showing the entire string you could show just the beginning and end. There are many use cases, like what I did [here](https://github.com/garu/Data-Printer/blob/master/lib/Data/Printer/Filter/Digest.pm#L11) to create a filter for (most) Digest::* modules on CPAN :)

Now, this is all very clear to me because I wrote it, so it's not really fair. Since it wasn't that clear to you, it means we could try and improve the documentation so others don't get confused.

Could you help? I'd love a Pull Request that would make the original documentation clearer - hopefully using less words than what I did explaining it here!

Thanks!
garu

**Comment by KES777** [comment]
If your module has '-class' why you need _data_printer? what is better to use and when?
Thank you very much for full answer, but things still looks complicated (( Maybe because example long and abstract and I should understand something more. Your module works at some different manner than I imagine. In any case I should dig  deepper before I will ready to improve something.

Lest chat on freenode, my nick is kes_, write a note please there. I was trying, but did not find you.

**Comment by garu** [comment]
Hey @KES777! I tried to reach you now on Freenode but it says `kes_: No such nick/channel`

The '-class' filter (or any other filter, for that matter) is meant so **you** can tell Data::Printer how to display some class. If you're creating a filter for a class (e.g. "My::Module") then whoever wrote My::Module has no say in that filter, it can be anything you want.

On the other hand, the `_data_printer` sub is meant so **the class author** can tell Data::Printer how to display the object (without having to add Data::Printer as a dependency). So if you're the author of My::Module, you can create a `sub _data_printer` so Data::Printer will know how it should display objects of the My::Module class, even if the user did not pre-defined any filters - the user doesn't even have to know about your module!

In short: the sub itself is created exactly the same, but:

* `sub _data_printer { ... }`
is placed inside a class (like My::Module) and means *"if someone tries to use Data::Printer to debug this object, this is what I want them to see"*.

* `use DDP filters => { 'My::Module' => sub { ... } };` (or `'-class' => sub { ... }`)
is something you write inside your program, if there is no `sub _data_printer` inside the My::Module, and/or if you did not like whatever Data::Printer wrote as a dump for that particular object.

Better now?

Also worth noticing they are completely optional, and most people use Data::Printer without ever having to learn about filters :)

btw I use an IRC proxy, so even if I don't reply right away, you can still send me questions and I'll answer them whenever I'm around - as long as you are around too! :D

**Comment by KES777** [comment]
>btw I use an IRC proxy, so even if I don't reply right away, you can still send me questions and I'll answer them whenever I'm around - as long as you are around too! :D

Maybe better to implement #100?

**Comment by KES777** [comment]
>Better now?

Yes. It is better. I have some suggestion how to improve that. Posted at #115 [here](https://github.com/garu/Data-Printer/issues/115#issuecomment-366436991)

Another question about filter. How to not dump some keys at hash?

Here is example how I do that with `Data::Dump::Filtered`:

```
my $filter =  sub {
        my( $obj, $cur ) =  @_;

        return { hide_keys => [ '_result_source', 'schema', 'CONFIG' ] };
};

Data::Dump::Filtered::dump_filtered( $some_data, $filter )

```

How to do same with `DDP`?

**Comment by KES777** [comment]
ping

**Comment by garu** [comment]
Hey @KES777! The version on master (soon to be 1.00 on CPAN, currently released as 0.99_X) let's you do this:

```perl
use DDP; p %hash, ignore_keys => [qw( _result_source schema CONFIG )];
```

Hope this helps! I'm adding more documentation for the new version here on github and making DEV releases on CPAN. Once they are through, I'll release 1.00.

I'm going to close this issue but if you have any other questions or concerns, please don't hesitate to ask and/or open new issues!

Thanks again for using Data::Printer and taking the time to review it. I really appreciate it!
